# Numerical integration in Python

The programming tasks in the next Application Exercise will require that you can evaluate (numerical) integrals in Python. This is rather straightforward, using the `quad()` function from the `scipy.integrate` package. 

Let's illustrate this with a simple example. You can easily evaluate
$$\int_0^1 x^2\,dx=\left[\frac{x^3}{3}\right]_0^1=\frac{1}{3}.$$
Next, let us use Python to calculate a numerical approximation of the same integral. It can be done like this:

In [1]:
# Import the quad() function from scipy.integrate
from scipy.integrate import quad

# Define the function we want to integrate
def f(x) :
    return x**2

# Then integrate it from 0 to 1 and print the result.
result = quad(f,0,1)
print(result)

(0.33333333333333337, 3.700743415417189e-15)


The output which came out as 
`(0.33333333333333337, 3.700743415417189e-15)`. 

The `quad()` function returns a tuple whose *first* entry is the calculated value while the second entry is an upper bound of the absolute error. In other words, the above result should be interpreted as:
$$\int_0^1 x^2\,dx=0.33333333333333337\pm 3.700743415417189\cdot 10^{-15}.$$

If we are not interested in the error estimate or (more likely) need to extract the calculated value for further calculations, we can of course access it using indexing (remember that the first entry has index `0`):


In [2]:
value = quad(f,0,1)[0]
print(value)

0.33333333333333337


You can of course pass built in functions to `quad()` as well. Let's calculate
$$\int_0^\pi \sin{x}\,dx\quad:$$

In [3]:
import numpy as np

value = quad(np.sin,0,np.pi)[0]
print(value)

2.0


## Anonymous functions
In our example above, we defined the function `f` before we passed it to `quad`. This is what we did:

In [4]:
def f(x) :
    return x**2

print(quad(f,0,1))

(0.33333333333333337, 3.700743415417189e-15)


Python also supports *anonymous functions* (called *lambda functions* in Python), which you can define on the fly as you provide them as argument to another function. To calculate the above integral using that method instead, we would do the following:

In [5]:
print(quad(lambda x : x**2,0,1))

(0.33333333333333337, 3.700743415417189e-15)


The `lambda x: <expression in x>` notation, saves you a little bit of work when you want to pass a simple function as an argument to another function and don't have any need for that function elsewhere in your code. This method doesn't assign any name to the function, which is why we call it an "anonymous" function.

## Revision of `for`-loops

Recall that `range(n)` will generate an iterable object containing the integers from `0` to `n-1`. This can be iterated over in a loop like this:

In [6]:
for k in range(4) :
    print(k)

0
1
2
3


If `m` and `n` are integers $m<n$, `range(m,n)` will give you the integers from `m` to `n-1`, i,e.

In [7]:
for k in range(2,4) :
    print(k)


2
3


So, let's say we want to calculate the sum
$$\sum_{k=1}^3 \int_0^1 x^k\,dx.$$
Note that the answer is
$$\int_0^1 x\,dx+\int_0^1 x^2\,dx+\int_0^1 x^3\,dx=\frac{1}{2}+\frac{1}{3}+\frac{1}{4}=\frac{13}{12}.$$
The following loop will do the trick:


In [8]:
s = 0
for k in range(1,4) :
    s = s+quad(lambda x: x**k,0,1)[0]  
    
print(s)

1.0833333333333335


Note that we extracted the first entry (index zero) from the result of quad to access the actual value.

### This should be all you need for the next AE, but feel free to experiment below:

In [ ]:
def Fpoly(f, n, x):
    T = ((1/np.pi)*quad(lambda t: f(t)*np.cos(0), -np.pi, np.pi)[0])/2
    for i in range(n):
        k = i+1
        ak = (1/np.pi)*quad(lambda t: f(t)*np.cos(k*t),-np.pi,np.pi)[0]
        bk = (1/np.pi)*quad(lambda t: f(t)*np.sin(k*t),-np.pi,np.pi)[0]
        T = T + ak*np.cos(k*x) + bk*np.sin(k*x)
    return T

def Fpoly(f,n,x) :
    # Supply the missing code such that the function follows the specifications.
    T = ((1/np.pi) * quad(lambda t : f(t)*np.cos(0), -np.pi, np.pi)[0])/2   # Let T be the first term a0/2 (use quad to calculate the integral).
    for k in range(1,n+1) :
        ak = quad(lambda t : f(t)*np.cos(k*t),-np.pi,np.pi)[0]/np.pi # Coefficient a_k
        bk = quad(lambda t : f(t)*np.sin(k*t),-np.pi,np.pi)[0]/np.pi # Coefficient b_k
        T = T + ak*np.cos(k*x) + bk*np.sin(k*x)
    return T

In [ ]:
def Fpolyc(f,c,n,x):
    T = ((1/c)*quad(lambda t: f(t)*np.cos((k*np.pi*t)/c),-c,c)[0])/2
    for i in range(n):
        k = i+1
        ak = (1/c)*quad(lambda t: f(t)*np.cos((k*np.pi*t)/c),-c,c)[0]
        bk = (1/c)*quad(lambda t: f(t)*np.sin((k*np.pi*t)/c),-c,c)[0]
        T = T + ak*np.cos((k*np.pi*x)/c) + bk*np.sin((k*np.pi*x)/c)
    return T

def Fpolyc(f,c,n,x) :
    # Supply the missing code such that the function follows the specifications.
    T = ((1/np.pi) * quad(lambda t : f(t)*np.cos(0), -c, c)[0])/2   # Let T be the first term a0/2 (use quad to calculate the integral).
    for k in range(1,n+1) :
        ak = quad(lambda t : f(t)*np.cos((k*np.pi*t)/c),-c,c)[0]/c # Coefficient a_k
        bk = quad(lambda t : f(t)*np.sin((k*np.pi*t)/c),-c,c)[0]/c # Coefficient b_k
        T = T + ak*np.cos((k*np.pi*x)/c) + bk*np.sin((k*np.pi*x)/c)
    return T